patterns/dxp/aspect_oriented_programming.md

```diff
@@ -0,0 +1,53 @@
+# Aspect Oriented Programming (AOP) guidelines
+
+> **NOTE:** This plugin was originally in metal-plugins but now lives in [liferay-portal](https://github.com/liferay/liferay-portal) so we can provide a better developing experience.
+
+AOP provides tools to improve control of what is executed before and after another method is invoked. You can find this component in the [`frontend-js-web`](https://github.com/liferay/liferay-portal/tree/c0c13433600398fed8768f539aa8212978f7409c/modules/apps/frontend-js/frontend-js-web) module.
+
+See more about Aspect Oriented Programming in [this article](https://medium.com/@kyuwoo.choi/sneak-peek-to-javascript-aop-16458f807842).
+
+## Prerequisites
+
+First of all the `frontend-js-web` dependency should be added to the `package.json` of the module in which we'll use AOP. Be sure that the version matches the one in [`frontend-js-web/bnd.bnd`](https://github.com/liferay/liferay-portal/blob/c0c13433600398fed8768f539aa8212978f7409c/modules/apps/frontend-js/frontend-js-web/bnd.bnd).
+
+```
+"dependencies": {
+	...
+	"frontend-js-web": "4.0.0",
+	...
+},
+```
+
+## Use case: Change the return value of an instance method invocation
+
+```javascript
+import {AOP} from `frontend-js-web`;
+
+const foo = new Foo();
+foo.bar(); // "bar"
+
+const _interceptedFoo = AOP.after(target => {
```

Comment:
We have an unused const here which is a bit fishy. What's this return value and what can you do with it?

We should probably want to document how the handle contains a `detach` method to clean up when the interception is no longer required (e.g. during the component destroy lifecycle)

Reply:
Done!